src/Workspaces/Core/Portable/CodeFixes/FixAllOccurrences/FixAllContext.DiagnosticProvider.cs

```diff
@@ -83,22 +84,23 @@ public abstract class DiagnosticProvider
                         case FixAllScope.Solution:
                             var projectsAndDiagnostics = ImmutableDictionary.CreateBuilder<Project, ImmutableArray<Diagnostic>>();
 
-                            var tasks = project.Solution.Projects.Select(async p => new
-                            {
-                                Project = p,
-                                Diagnostics = await fixAllContext.GetProjectDiagnosticsAsync(p).ConfigureAwait(false)
-                            }).ToArray();
-
-                            await Task.WhenAll(tasks).ConfigureAwait(false);
-
-                            foreach (var task in tasks)
-                            {
-                                var projectAndDiagnostics = await task.ConfigureAwait(false);
-                                if (projectAndDiagnostics.Diagnostics.Any())
+                            await ProducerConsumer<(Project project, ImmutableArray<Diagnostic> diagnostics)>.RunParallelAsync(
+                                source: project.Solution.Projects,
+                                produceItems: static async (project, callback, args, cancellationToken) =>
+                                {
+                                    var diagnostics = await args.fixAllContext.GetProjectDiagnosticsAsync(project).ConfigureAwait(false);
+                                    callback((project, diagnostics));
+                                },
+                                consumeItems: static async (results, args, cancellationToken) =>
                                 {
-                                    projectsAndDiagnostics[projectAndDiagnostics.Project] = projectAndDiagnostics.Diagnostics;
-                                }
-                            }
+                                    await foreach (var (project, diagnostics) in results)
+                                    {
+                                        if (diagnostics.Any())
```

Comment:
>if (diagnostics.Any()) [](http://example.com/codeflow?start=40&length=22)

nit: might make more sense to filter this on the producer side

Reply:
agreed.  can do in followup.